azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageMgmtCommonDaoImpl.java

```diff
@@ -0,0 +1,230 @@
+ * version and its associated metadata.
+ */
+public class ImageMgmtCommonDaoImpl implements ImageMgmtCommonDao {
+
+  private static final Logger log = LoggerFactory.getLogger(ImageRampupDaoImpl.class);
+
+  private final DatabaseOperator databaseOperator;
+
+  private final ImageTypeDao imageTypeDao;
+
+  private final ImageVersionDao imageVersionDao;
+
+  private final ImageRampupDao imageRampupDao;
+
+  // Make sure correct rampups are deleted based on the image type and version id
+  private static final String DELETE_IMAGE_RAMPUP_BY_VERSION_ID_QUERY = "delete from image_rampup "
+      + "where plan_id in ( ${plan_ids} )";
+
+  // Make sure correct rampup plan is deleted for the specified image type
+  private static final String DELETE_IMAGE_RAMPUP_PLAN_BY_PLAN_ID_QUERY = "delete from "
+      + "image_rampup_plan where id in ( ${plan_ids} )";
+
+  // Make sure correct image version is deleted based on image type and version id
+  private static final String DELETE_IMAGE_VERSION_BY_IMAGE_TYPE_AND_VERSION_ID_QUERY = "delete "
+      + "iv from image_versions iv inner join image_types it on it.id = iv.type_id where iv.id = "
+      + "? and  lower(it.name) = ? ";
+
+  @Inject
+  public ImageMgmtCommonDaoImpl(final DatabaseOperator databaseOperator,
+      final ImageTypeDao imageTypeDao,
+      final ImageVersionDao imageVersionDao,
+      final ImageRampupDao imageRampupDao) {
+    this.databaseOperator = databaseOperator;
+    this.imageTypeDao = imageTypeDao;
+    this.imageVersionDao = imageVersionDao;
+    this.imageRampupDao = imageRampupDao;
+  }
+
+  @Override
+  public DeleteResponse deleteImageVersion(final String imageTypeName, final int versionId,
+      final Boolean forceDelete) throws ImageMgmtException {
+    final DeleteResponse deleteResponse = new DeleteResponse();
+    // Get the image version for the image type and id. Throw exception if not exits.
+    final Optional<ImageVersion> imageVersion = this.imageVersionDao.getImageVersion(
+        imageTypeName, versionId);
+    if (!imageVersion.isPresent()) {
+      deleteResponse.setErrorCode(ErrorCode.NOT_FOUND);
+      deleteResponse.setMessage(String.format("The version id: %d for image type: %s does not "
+          + "exist.", versionId, imageTypeName));
+      return deleteResponse;
+    }
+    // Get the rampup plan and rampup details for the image type and version id.
+    final List<ImageRampupPlan> imageRampupPlans = this.imageRampupDao
+        .getImageRampupPlans(imageTypeName,
+            versionId);
+    // Collect the plan ids and delete the image rampup and rampup plan using that.
+    final List<Integer> planIds = new ArrayList<>();
+    if (!CollectionUtils.isEmpty(imageRampupPlans)) {
+      for (final ImageRampupPlan imageRampupPlan : imageRampupPlans) {
+        planIds.add(imageRampupPlan.getId());
+      }
+      // If the version is used in any of the rampup plans throw error with the rampup plan details.
```

Comment:
Here it means only active rampup plans right? If so, then there should be only single plan id coming. If not then what is the point of deleting inactive plans? Another thing to think about is shall we just do soft delete for image version? If version is marked for soft delete then it should not be allowed to be picked up by any ramp up plan.

Reply:
The idea to completely delete image version and associated metadata which includes image rampup plan and rampup details. Inactive rampup plans are of no use as we allow update only on active rampup plan. Hence, while deleting rampup plans inacitve plans are also considered. 

Regarding the soft delete of image versions we have to put many if and checks in the code. For example, if the version is marked as soft delete, what will happen to the rampup plan and rampup details. They also need to be marked as soft delete. Also we need to put some conditions in the code as well. Also we have foreign key constraints on all these tables. Moreover, the version cleanup can be leveraged to clean up of unused data as well.

The delete image version API ensures that if the version is in NEW/ACTIVE state or there are associated rampup plans, the delete will fail and display the usage details. Only by specifying forceDelete flag to "true" we should be able to forcefully delete the image version metadata.